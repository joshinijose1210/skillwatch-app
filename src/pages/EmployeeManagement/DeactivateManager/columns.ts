import { DeactivateManagerFormatter, ReporteeNameFormatter } from '@components';
import { TableColumnConfig } from '@medly-components/core';

export const DeactivateManagerSchema: TableColumnConfig[] = [
    {
        title: 'Reportee Name',
        field: '',
        component: ReporteeNameFormatter
    },
    {
        title: 'Change Manager To',
        field: '',
        component: DeactivateManagerFormatter
    }
];
