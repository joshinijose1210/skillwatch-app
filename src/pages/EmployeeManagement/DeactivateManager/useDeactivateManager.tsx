import { ErrorType } from '@common';
import { routeConstants } from '@constants';
import { addToast } from '@medly-components/core';
import { useAppSelector } from '@slice';
import { useEditEmployeeMutation, usePostDeactivateManagerMutation } from '@slice/services';
import { setManagerList } from '@slice/updateManager';
import { useEffect, useMemo, useState } from 'react';
import { useDispatch } from 'react-redux';
import { useLocation, useNavigate } from 'react-router-dom';

const useDeactivateManager = () => {
    const dispatch = useDispatch(),
        navigate = useNavigate(),
        { employeeDetails, managerData, isLoading, heading } = useLocation().state as any,
        { managerUpdateDataList } = useAppSelector(state => state.updateManager);
    const [activePage, setActivePage] = useState(1),
        [columnData, setColumnData] = useState([]);

    const [editEmployee, { isLoading: editEmployeeLoading, error: editEmployeeError, isSuccess: editEmployeeSuccess }] =
            useEditEmployeeMutation(),
        [postDeactivateManager, { isLoading: editManagerLoading }] = usePostDeactivateManagerMutation();
    const isDisabled = useMemo(
        () => managerUpdateDataList.some(data => data.newManagerId === '' || data.currentManagerId === data.newManagerId),
        [managerUpdateDataList]
    );

    const handlePageChange = (page: number) => {
            setActivePage(page);
        },
        handelSubmit = () => {
            if (isDisabled) return;
            const updatedDetails = employeeDetails;
            employeeDetails.secondManagerId === -1 && delete updatedDetails['secondManagerId'];
            if (heading === 'Change Manager') {
                updatedDetails.status = false;
            }
            editEmployee({ path: employeeDetails.id, data: { updatedDetails } });
            postDeactivateManager({
                managerUpdateDataList
            });
        };

    useEffect(() => {
        if (managerData && managerData.managerId) {
            setColumnData(
                managerData.reporteesData.map((element: any) => ({
                    ...element,
                    showDotIndication: heading.toLowerCase() !== 'change role',
                    currentManagerId: managerData.managerId
                }))
            );
            const newManagerData = managerData.reporteesData.map((element: any) => ({
                currentManagerId: managerData.managerId,
                employeeId: element.id,
                newManagerId: ''
            }));
            dispatch(setManagerList(newManagerData));
        }
    }, [managerData, dispatch]);

    useEffect(() => {
        if (typeof editEmployeeError === 'object') {
            const errorObj = editEmployeeError as ErrorType;
            addToast({
                variant: 'error',
                header: errorObj.data.errorMessage || 'Something went wrong',
                timer: 3000
            });
        } else if (editEmployeeSuccess === true) {
            navigate(`${routeConstants.employeeManagement}/${1}`);
            addToast({
                variant: 'success',
                header: 'Employee updated successfully',
                timer: 3000
            });
        }
    }, [dispatch, editEmployeeError, editEmployeeSuccess, managerUpdateDataList, navigate]);

    return {
        handlePageChange,
        columnData,
        activePage,
        handelSubmit,
        isLoading,
        isDisabled,
        editManagerLoading,
        editEmployeeLoading,
        heading
    };
};

export default useDeactivateManager;
