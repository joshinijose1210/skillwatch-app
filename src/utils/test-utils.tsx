import { ToastContainer } from '@medly-components/core';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import { store as rtkStore, setupStore } from '@slice';
import { RenderOptions, RenderResult, render } from '@testing-library/react';
import { defaultTheme } from '@theme';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import reduxMockStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import { userMockState } from '../mocks/userMockState';
import { initialState } from '../slice';

interface CustomOptions extends RenderOptions {
    initialState?: any;
    initialEntries?: any;
}

interface WithThemeProviderProps {
    children?: React.ReactNode;
}

const mockStore = reduxMockStore(),
    store = mockStore({
        ...initialState,
        user: { ...initialState.user, groups: ['admin'] }
    });

const WithThemeProvider: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <ThemeProvider theme={defaultTheme}>
        <>
            <ToastContainer position="top-end" />
            {children}
        </>
    </ThemeProvider>
);

const WithStore: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <WithThemeProvider>{children}</WithThemeProvider>
    </Provider>
);

const WithRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <MemoryRouter>
        <WithThemeProvider>{children}</WithThemeProvider>
    </MemoryRouter>
);

const WithStoreAndRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <MemoryRouter>
            <WithThemeProvider>{children}</WithThemeProvider>
        </MemoryRouter>
    </Provider>
);

export const RTKRender = (ui: React.ReactElement, { initialState = userMockState, ...renderOptions }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);

    const RtkWrapper: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
        <Provider store={store}>
            <MemoryRouter initialEntries={renderOptions.initialEntries}>
                <WithThemeProvider>{children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
    return { rtkStore, ...render(ui, { wrapper: RtkWrapper, ...renderOptions }) };
};
const customRender = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithThemeProvider, ...options });

export const renderWithStore = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStore, ...options });

export const renderWithRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithRouter, ...options });

export const renderWithStoreAndRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStoreAndRouter, ...options });

export const wrapper = (props: { children: React.ReactNode }, { initialState = userMockState }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);
    return (
        <Provider store={store}>
            <MemoryRouter>
                <WithThemeProvider>{props.children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
};

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render, store };
