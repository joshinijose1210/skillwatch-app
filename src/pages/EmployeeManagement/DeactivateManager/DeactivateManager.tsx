import { Loader, PageContent } from '@components';
import { CustomTable } from '@components/reusableComponents/CustomTable/CustomTable';
import ListHeader from '@components/reusableComponents/ListHeader';
import { Button } from '@medly-components/core';
import { ButtonDiv } from '@pages/RolesAndPermissions/RoleForm/RoleForm.styled';
import { FC } from 'react';
import { DeactivateManagerSchema } from './columns';

import useDeactivateManager from './useDeactivateManager';
const DeactivateManager: FC = () => {
    const {
        activePage,
        handlePageChange,
        isDisabled,
        columnData,
        handelSubmit,
        isLoading,
        editEmployeeLoading,
        editManagerLoading,
        heading
    } = useDeactivateManager();

    if (isLoading) {
        return <Loader />;
    } else
        return (
            <PageContent>
                <ListHeader title={heading || 'Change Manager'} />
                <CustomTable
                    count={columnData.length}
                    isLoading={isLoading}
                    data={columnData}
                    itemsPerPage={100}
                    columns={DeactivateManagerSchema}
                    handlePageChange={handlePageChange}
                    activePage={activePage}
                />
                <ButtonDiv>
                    <Button
                        data-testid="saveBtn"
                        disabled={isDisabled}
                        isLoading={editEmployeeLoading || editManagerLoading}
                        onClick={handelSubmit}
                    >
                        Save
                    </Button>
                </ButtonDiv>
            </PageContent>
        );
};
export default DeactivateManager;
