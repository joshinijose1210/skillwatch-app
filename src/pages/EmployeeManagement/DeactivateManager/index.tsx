import DeactivateManager from './DeactivateManager';
export default DeactivateManager;
