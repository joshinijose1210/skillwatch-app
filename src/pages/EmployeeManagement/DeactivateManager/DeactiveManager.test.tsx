import { RTKRender, cleanup, fireEvent, screen, waitFor } from '@utils/test-utils';
import DeactivateManager from '.';
import MockAdapter from 'axios-mock-adapter';
import * as reactRouterDom from 'react-router-dom';
import { addEmployeeMockData } from '@mocks/addEmployee';
import { apiInstance } from '@utils';
import { Loader } from '@components';
const mock = new MockAdapter(apiInstance);

const mockNavigate = jest.fn();

const useLocationMock = jest.spyOn(reactRouterDom, 'useLocation');

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: jest.fn(),
    useNavigation: mockNavigate
}));

beforeEach(() => {
    useLocationMock.mockReturnValue({ ...addEmployeeMockData.LocationMockData, state: addEmployeeMockData.navigationRoleState.state });
    mock.reset();
});

afterEach(cleanup);

describe('Deactive Manager component', () => {
    it('should take snapshot', () => {
        useLocationMock.mockReturnValue({
            ...addEmployeeMockData.LocationMockData,
            state: { isLoading: false }
        });
        const { container } = RTKRender(<DeactivateManager />);
        expect(container).toMatchSnapshot();
    });

    it('should show loader properly when page is loading', () => {
        useLocationMock.mockReturnValue({
            ...addEmployeeMockData.LocationMockData,
            state: { isLoading: true }
        });
        const { container } = RTKRender(<DeactivateManager />);
        const { container: loaderContainer } = RTKRender(<Loader />);
        expect(container.innerHTML).toMatch(loaderContainer.innerHTML);
    });

    it('should render properly', async () => {
        const { url, someData } = addEmployeeMockData.managers.get;
        mock.onGet(url).replyOnce(200, someData);
        RTKRender(<DeactivateManager />);
        const deactiveManager = await screen.findByTestId<HTMLInputElement>('deactiveManager');
        expect(screen.getByText('Change Role')).toBeInTheDocument();
        expect(screen.getByText('A S (SR006)')).toBeInTheDocument();
        expect(screen.getByTestId('saveBtn')).toBeDisabled();
        expect(deactiveManager).toBeInTheDocument();
        expect(deactiveManager.value).toBe('R P (SR002)');
    });

    it('should change manager properly in change role', async () => {
        const { url, someData } = addEmployeeMockData.managers.get;
        mock.onGet(url).replyOnce(200, someData);
        mock.onPut().replyOnce(200, { refresh_token: '' });
        RTKRender(<DeactivateManager />);
        expect(screen.getByText('Change Role')).toBeInTheDocument();
        const managerDropDown = screen.getByTestId('deactiveManager');
        const saveBtn = screen.getByTestId('saveBtn');
        await waitFor(() => expect(managerDropDown).toHaveValue('R P (SR002)'), { timeout: 3000 });
        expect(saveBtn).toBeDisabled();
        fireEvent.change(managerDropDown, { target: { value: 'R S (SR002)' } });
        await waitFor(() => expect(saveBtn).toBeEnabled(), { timeout: 3000 });
        fireEvent.click(saveBtn);
        expect(await screen.findByText('Employee updated successfully')).toBeInTheDocument();
    });

    it('should give an error when api fails', async () => {
        const { url, someData } = addEmployeeMockData.managers.get;
        mock.onGet(url).replyOnce(200, someData);
        mock.onPut('https://dummy-api.example.com/api/employees/').replyOnce(400, { errorMessage: 'Something went wrong' });
        mock.onPut('https://dummy-api.example.com/api/employees/2').replyOnce(400, { errorMessage: 'Something went wrong' });
        RTKRender(<DeactivateManager />);
        expect(screen.getByText('Change Role')).toBeInTheDocument();

        const managerDropDown = screen.getByTestId('deactiveManager');
        const saveBtn = screen.getByTestId('saveBtn');
        await waitFor(() => expect(managerDropDown).toHaveValue('R P (SR002)'), { timeout: 3000 });
        expect(saveBtn).toBeDisabled();
        fireEvent.change(managerDropDown, { target: { value: 'R S (SR002)' } });
        await waitFor(() => expect(saveBtn).toBeEnabled(), { timeout: 3000 });
        fireEvent.click(saveBtn);
        expect(await screen.findByText('Something went wrong')).toBeInTheDocument();
    });

    it('should test for change manager in deactive manager', async () => {
        const { url, someData } = addEmployeeMockData.managers.get;
        mock.onGet(url).replyOnce(200, someData);
        mock.onPut().replyOnce(200, { refresh_token: '' });
        useLocationMock.mockReturnValue({
            ...addEmployeeMockData.LocationMockData,
            state: { ...addEmployeeMockData.navigationRoleState.state, heading: 'Change Manager' }
        });
        RTKRender(<DeactivateManager />);
        expect(screen.getByText('Change Manager')).toBeInTheDocument();
        const managerDropDown = screen.getByTestId('deactiveManager');
        const saveBtn = screen.getByTestId('saveBtn');
        await waitFor(() => expect(managerDropDown).toHaveValue('R P (SR002)'), { timeout: 3000 });
        expect(saveBtn).toBeDisabled();
        fireEvent.change(managerDropDown, { target: { value: 'R S (SR002)' } });
        await waitFor(() => expect(saveBtn).toBeEnabled(), { timeout: 3000 });
        fireEvent.click(saveBtn);
        expect((await screen.findAllByText('Employee updated successfully'))[0]).toBeInTheDocument();
    });
});
