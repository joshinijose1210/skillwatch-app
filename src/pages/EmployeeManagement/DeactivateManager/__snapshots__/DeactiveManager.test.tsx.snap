// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deactive Manager component should take snapshot 1`] = `
<div>
  <main
    class="sc-1ehavxz-0 dDyhBJ"
  >
    <div
      class="sc-h70dbj-0 fQmYvl row expanded space-between align-center marginBottom-2"
    >
      <div
        class="sc-w8e7z6-2"
      >
        <h3
          class="sc-1ntgl3l-0 kcLBOi sc-w8e7z6-0 ibnmt"
          textalign="initial"
          textcolor="inherit"
          textvariant="h3"
          textweight="Medium"
        >
          Change Manager
        </h3>
      </div>
    </div>
    <div
      class="sc-8tbhpn-3"
    >
      <div
        class="sc-1msmxxr-0 iWEzjW"
      >
        Change Manager To
      </div>
      <table
        actions=""
        class="sc-1msmxxr-1 jwFsPF sc-8tbhpn-2 fqOFEu"
        columns="[object Object],[object Object]"
        defaultactivepage="1"
        defaultsortorder="desc"
        itemsperpage="10"
        rowclickdisablekey=""
        tabindex="-1"
        totalitems="0"
      >
        <thead
          class="sc-uxzsxp-0 haNltq"
        >
          <tr
            class="sc-1jr91bg-0 hJzMcO"
            gridtemplatecolumns=" minmax(100px, 1fr) minmax(100px, 1fr)"
            style="grid-template-columns: minmax(100px, 1fr) minmax(100px, 1fr);"
          >
            <th
              class="sc-hjmm4m-0 iOyNiA"
              tablesize="M"
            >
              <button
                class="sc-hjmm4m-2 kpjakb"
                tablesize="M"
              >
                <span
                  class="sc-1ntgl3l-0 ghpJwn"
                  textalign="initial"
                  textcolor="inherit"
                >
                  Reportee Name
                </span>
              </button>
              <span
                class="sc-hjmm4m-1 cuEDpa"
              />
            </th>
            <th
              class="sc-hjmm4m-0 iOyNiA"
              tablesize="M"
            >
              <button
                class="sc-hjmm4m-2 kpjakb"
                tablesize="M"
              >
                <span
                  class="sc-1ntgl3l-0 ghpJwn"
                  textalign="initial"
                  textcolor="inherit"
                >
                  Change Manager To
                </span>
              </button>
              <span
                class="sc-hjmm4m-1 cuEDpa"
              />
            </th>
          </tr>
        </thead>
        <tbody
          class="sc-c7cniy-0 kShdfA"
        >
          <tr
            class="sc-1yxujaa-0 sc-1yxujaa-2 cfQtmZ gtGPdU"
            gridtemplatecolumns=" minmax(100px, 1fr) minmax(100px, 1fr)"
            style="grid-template-columns: minmax(100px, 1fr) minmax(100px, 1fr);"
          >
            <td
              class="sc-1yxujaa-1 hmPqUc"
              tablesize="M"
              width="0"
            >
              No result
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="sc-1o9mfd9-2 gLsThh"
    >
      <button
        class="sc-1k2druj-0 eLgfbc"
        data-testid="saveBtn"
        edges="square"
        type="button"
        variant="solid"
      >
        <span
          class="sc-1ntgl3l-0 eOsHHD"
          textalign="initial"
          textcolor="inherit"
          textvariant="button1"
        >
          Save
        </span>
      </button>
    </div>
  </main>
</div>
`;
